### PR TITLE
Changing epochs

### PR DESCRIPTION
**Pull request by talekhinezh**
* Implement epoch changes occurring after some given number of views in a current epoch (uses the same validator set across every epoch though)
* Refactored consensus tests to support epochs and epochs tests

### CODE REVIEW

**Review by Sajjon — APPROVED** [review]
Much moar happy with this naming :)

File: radixdlt/src/main/java/com/radixdlt/consensus/EpochChange.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import com.radixdlt.consensus.validators.ValidatorSet;
+import java.util.Objects;
+
+/**
+ * An epoch change message to consensus
+ */
+public final class EpochChange {
+	private final VertexMetadata ancestor;
+	private final ValidatorSet validatorSet;
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/EpochChange.java:28]: It's not clear here whether `validatorSet` represents the next or the previous validator set

File: radixdlt/src/main/java/com/radixdlt/consensus/EpochManager.java [hunk 2]
```diff
@@ -18,80 +18,107 @@
...
+		final long nextEpoch = genesisVertex.getEpoch();
 
-		ProposerElection proposerElection = proposerElectionFactory.create(validatorSet);
-		log.info("NEXT_EPOCH: ProposerElection: {}", proposerElection);
+		// Sanity check
+		if (nextEpoch <= this.currentEpoch) {
+			throw new IllegalStateException("Epoch change has already occurred: " + epochChange);
+		}
+
+		counters.set(CounterType.EPOCHS_EPOCH, nextEpoch);
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/EpochManager.java:101]: "counters"

File: radixdlt/src/main/java/com/radixdlt/consensus/SyncedStateComputer.java [hunk 3]
```diff
@@ -44,13 +44,24 @@
...
+	/**
+	 * Computes whether commit of this vertex will cause an epoch change.
+	 *
+	 * TODO: Return the entire computation result
+	 * TODO: Transform into the prepare/execute stage
+	 *
+	 * @param vertex the vertex to compute
+	 * @return true, if commit of this vertex will cause an epoch change, false otherwise
+	 */
+	boolean compute(Vertex vertex);
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/SyncedStateComputer.java:64]: `compute` does not convey the actual meaning of `whether commit of this vertex will cause an epoch change`

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java [hunk 4]
```diff
@@ -386,26 +394,37 @@ private boolean addQC(QuorumCertificate qc) {
...
-	private void insertVertexInternal(Vertex vertex) throws VertexInsertionException {
+	private VertexMetadata insertVertexInternal(Vertex vertex) throws VertexInsertionException {
 		if (!vertices.containsKey(vertex.getParentId())) {
 			throw new MissingParentException(vertex.getParentId());
 		}
 
+		final Vertex vertexToUse;
+		if (vertex.getParentMetadata().isEndOfEpoch()) {
+			// TODO: Check if current proposal is actually an empty client atom
+			vertexToUse = new Vertex(vertex.getEpoch(), vertex.getQC(), vertex.getView(), null);
```
Comment (flotothemoon) [inline radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java:405]: Why do we need to force drop the client atom here?

**Review by flotothemoon — COMMENTED** [review]
Looks good for the most part, will have to take another look.

File: radixdlt/src/integration/java/com/radixdlt/consensus/simulation/invariants/bft/LivenessInvariant.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+public class LivenessInvariant implements TestInvariant {
+	private final long duration;
+	private final TimeUnit timeUnit;
+
+	public LivenessInvariant(long duration, TimeUnit timeUnit) {
+		this.duration = duration;
+		this.timeUnit = timeUnit;
+	}
+
+	private Comparator<VertexMetadata> vertexMetadataComparator =
```
Comment (clstrfsck) [inline radixdlt/src/integration/java/com/radixdlt/consensus/simulation/invariants/bft/LivenessInvariant.java:45]: Would be easier to see intent as:
```
private Comparator<VertexMetadata> vertexMetadataComparator =
	Comparator.comparingLong(VertexMetadata::getEpoch).thenComparing(VertexMetadata::getView);
```

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexMetadata.java [hunk 6]
```diff
@@ -46,35 +50,56 @@
 	@DsonOutput(Output.ALL)
 	private final long stateVersion;
 
+	@JsonProperty("isEndOfEpoch")
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/VertexMetadata.java]: We've been fairly consistently using snake_case on these.

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java (hunk 4 above)
Comment (clstrfsck) [reply radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java:405]: Wouldn't it be better to decline to vote on an epoch change proposal with an atom at the BFT level, and have this as failure case at this level?

Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java:405]: @clstrfsck yes, that would be better but saving that for the next PR.

File: radixdlt/src/integration/java/com/radixdlt/consensus/simulation/invariants/bft/LivenessInvariant.java (hunk 5 above)
Comment (talekhinezh) [reply radixdlt/src/integration/java/com/radixdlt/consensus/simulation/invariants/bft/LivenessInvariant.java:45]: yes

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexMetadata.java (hunk 6 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/VertexMetadata.java]: good catch

File: radixdlt/src/main/java/com/radixdlt/consensus/EpochChange.java (hunk 1 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/EpochChange.java:28]: I think it's clear enough given that there is an ancestor which refers to the "past".

File: radixdlt/src/main/java/com/radixdlt/consensus/SyncedStateComputer.java (hunk 3 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/SyncedStateComputer.java:64]: well, it is computing whether an epoch change will occur?

File: radixdlt/src/integration/java/com/radixdlt/consensus/simulation/invariants/bft/LivenessInvariant.java (hunk 5 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/consensus/simulation/invariants/bft/LivenessInvariant.java:45]: Did you intent to change this?  Comment suggests that you did, but I don't see any changes.

File: radixdlt/src/main/java/com/radixdlt/consensus/EpochManager.java [hunk 7]
```diff
@@ -55,15 +57,16 @@
 	private final ECKeyPair selfKey;
 	private final SystemCounters counters;
 	private final Hasher hasher;
+	private final ScheduledTimeoutSender scheduledTimeoutSender;
 
 	private long currentEpoch;
 	private VertexStore vertexStore;
 	private BFTEventProcessor eventProcessor = EMPTY_PROCESSOR;
 
-	@Inject
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/EpochManager.java:63]: "Named" annotation could be removed below also.

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexMetadata.java (hunk 6 above)
Comment (clstrfsck) [reply radixdlt/src/main/java/com/radixdlt/consensus/VertexMetadata.java]: Doesn't seem to be resolved?

File: radixdlt/src/integration/java/com/radixdlt/consensus/simulation/invariants/bft/LivenessInvariant.java (hunk 5 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/consensus/simulation/invariants/bft/LivenessInvariant.java:45]: NVM.  Looking at wrong commit.

File: radixdlt/src/test/java/com/radixdlt/consensus/AddressBookValidatorSetProviderTest.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
+package com.radixdlt.consensus;
```
Comment (clstrfsck) [inline radixdlt/src/test/java/com/radixdlt/consensus/AddressBookValidatorSetProviderTest.java:18]: Missing header comment.

File: radixdlt/src/test/java/com/radixdlt/consensus/EpochManagerTest.java [hunk 9]
```diff
@@ -10,6 +10,7 @@
 import com.radixdlt.consensus.VertexStore.GetVerticesRequest;
```
Comment (clstrfsck) [inline radixdlt/src/test/java/com/radixdlt/consensus/EpochManagerTest.java:27]: Missing header comment here too.

**Review by clstrfsck — APPROVED** [review]
Looks good once missing header comments are fixed.

**Pull request merged**